### PR TITLE
Add client error message on bootstrap rejected because of blacklist/whitelist

### LINKED ISSUE #3292: Better error message on bootstrap rejected because of blacklist/whitelist
```
2022-12-01T17:38:55.403601Z  INFO massa_node: Node version : TEST.17.0
2022-12-01T17:38:55.435090Z  INFO massa_bootstrap::client: Start bootstrapping from 195.134.167.226:31245
2022-12-01T17:38:55.435808Z  WARN massa_bootstrap::client: Error while bootstrapping: io error: early eof
```

"early eof" is not very informative of the problem which is lack of whitelisting.

Maybe do it like we do elsewhere: send an error message from the server if there are slots available, otherwise just close

### PR DESCRIPTION
**Pull request by aoudiamoncef**
* [x] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [x] add logs allowing easy debugging in case the changes caused problems
* [x] if the API has changed, update the API specification

closes #3292

### CODE REVIEW

**Comment by aoudiamoncef** [comment]
@Eitu33 

This was quickly tested locally and with two physical nodes

Results with a simple telnet calls:

blacklist:
```shell
telnet 192.168.1.190 31245
Trying 192.168.1.190...
Connected to 192.168.1.190.
Escape character is '^]'.
?q??-?h?,Gs?S%?`???9????'?Q[?g????T?T"y+>F?^ox?R????{??" IP 192.168.1.190 is blacklisted.Connection closed by foreign host.
```

not in whitelist:
```shell
telnet 192.168.1.190 31245
Trying 192.168.1.190...
Connected to 192.168.1.190.
Escape character is '^]'.
??S?.
_???;R??M?o?Z?????c?	?t?f??)???cy?a?k?a5?N?f??A?A whitelist exists and the IP 192.168.1.190 is not whitelisted.Connection closed by foreign host.
```

File: massa-bootstrap/src/server.rs
Comment (Eitu33) [inline massa-bootstrap/src/server.rs:334]: Creating a new `BootstrapServerBinder` for one message is not optimal, please take it as a function argument. This will also avoid a lot of duplicated code.

Comment (aoudiamoncef) [reply massa-bootstrap/src/server.rs:334]: Before calling this function, there is no instance of it. It could be not used at all, this why it's created only if needed. 
It's implemented that way in the rest of the `run`function.

Comment (aoudiamoncef) [reply massa-bootstrap/src/server.rs:334]: May be we should create a single instance and use it in `run` function ?

Comment (Eitu33) [reply massa-bootstrap/src/server.rs:334]: I see, the code is still seriously duplicated though. The only difference between the two cases are the error messages, the binder creation code can be written only once and the conditional statements should only take care of the messages content.

Comment (aoudiamoncef) [reply massa-bootstrap/src/server.rs:334]: i'll refactor this ASAP. We could share a single instance of `BootstrapServerBinder`:

```rust
let (mut server, remote_addr) = if res_connection.is_ok() {
                        let (dplx, remote_addr) = res_connection.unwrap();
                        let server = BootstrapServerBinder::new(dplx, self.keypair.clone(), self.bootstrap_config.max_bytes_read_write, self.bootstrap_config.max_bootstrap_message_size, self.bootstrap_config.thread_count, self.bootstrap_config.max_datastore_key_length, self.bootstrap_config.randomness_size_bytes, self.bootstrap_config.consensus_bootstrap_part_size);
                        // we didn't test whether Ip is allowed
```

Comment (aoudiamoncef) [reply massa-bootstrap/src/server.rs]: done

File: massa-bootstrap/src/server.rs
Comment (Eitu33) [inline massa-bootstrap/src/server.rs:192]: Indentation seems wrong here, maybe you didn't run fmt but its probably the formatter bugging out because of the select or the test scopes. If that's the case you have to do it manually.

Comment (Eitu33) [reply massa-bootstrap/src/server.rs:192]: Otherwise lgtm, will merge when this is fixed 👍

Comment (Eitu33) [reply massa-bootstrap/src/server.rs:192]: (And CI is passing)

File: massa-bootstrap/src/server.rs
Comment (Ben-PH) [inline massa-bootstrap/src/server.rs:217]: This might be a nice place to use let-else:

```suggestion
                            let Ok(res) = self.is_ip_allowed(remote_addr, server, &whitelist, &blacklist).await else {
                                continue;
                            };
```

Comment (Ben-PH) [reply massa-bootstrap/src/server.rs:217]: (wrote this out manually in github. Not sure if the `;` on line 213 is needed)

Comment (aoudiamoncef) [reply massa-bootstrap/src/server.rs:217]:
```
                        let (mut server, remote_addr) = if let Ok(res) = res_connection {
                            let server = BootstrapServerBinder::new(
                                res.0,
                                self.keypair.clone(),
                                self.bootstrap_config.max_bytes_read_write,
                                self.bootstrap_config.max_bootstrap_message_size,
                                self.bootstrap_config.thread_count,
                                self.bootstrap_config.max_datastore_key_length,
                                self.bootstrap_config.randomness_size_bytes,
                                self.bootstrap_config.consensus_bootstrap_part_size);
                        // we didn't test whether Ip is allowed
                        #[cfg(test)]
                        {
                            (server, res.1)
                        }

                        #[cfg(not(test))]
                        {
                            // check whether incoming peer IP is allowed or return an error which is ignored
                            let Ok(res) = self.is_ip_allowed(res.1, server, &whitelist, &blacklist).await else {
                                continue;
                            };
                            res
                        }
                    } else {
                        continue
                    };
```

@Eitu33 @Ben-PH is it more readable like this ? (not sure)

Comment (aoudiamoncef) [reply massa-bootstrap/src/server.rs]: done

File: massa-bootstrap/src/server.rs
Comment (aoudiamoncef) [reply massa-bootstrap/src/server.rs]: done

File: massa-bootstrap/src/server.rs
Comment (Ben-PH) [reply massa-bootstrap/src/server.rs]: I think so, yeah. The dropping of the `unwrap` is nice in the old line 193 is a nice extra.

Comment (aoudiamoncef) [reply massa-bootstrap/src/server.rs]: IMHO, it's already done ✅

**Comment by Eitu33** [comment]
We can merge without gas costs checks passing for now but the ubuntu CI fails also on unused, please fix before we merge

**Comment by aoudiamoncef** [comment]
@Eitu33 done 💯

**Review by Eitu33 — APPROVED** [review]

**Pull request merged**
